python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,433 @@
+import copy
+import enum
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.ml import Checkpoint
+from ray.tune.result import NODE_IP
+from ray.util import PublicAPI
+from ray.util.annotations import DeveloperAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+class CheckpointStorage(enum.Enum):
+    MEMORY = enum.auto()
+    PERSISTENT = enum.auto()
+
+
+@DeveloperAPI
+class TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager.
+
+    This class is used to track checkpoints generated by trainables and trainers in
+    order to add metadata (e.g. the result, or the node where it has been created)
+    and for bookkeeping purposes.
+
+    The data can be an object, a checkpoint directory, or a future to either. Because
+    we can't know if it's data or a directory from a future, this class expects
+    a ``storage_mode`` that makes the data type explicit.
+
+    The passed metrics can be used to compare performance of different checkpoints.
+    The ``checkpoint_id`` is passed as an alternative to be able to order
+    checkpoints in time.
+
+    Args:
+        dir_or_data: Checkpoint directory, checkpoint data, or a future to either.
+        storage_mode: Either MEMORY or PERSISTENT.
+        checkpoint_id: Checkpoint number. Will be used to determine checkpoint order
+            if metrics are not available. Usually this should be monotonically
+            increasing for each tracked checkpoint.
+        metrics: Observed metrics for this checkpoint. This is used to determine
+            the value of the ``checkpoint_score_attr``.
+        node_ip: IP of the node where the checkpoint was generated. Defaults
+            to the current node.
+    """
+
+    def __init__(
+        self,
+        dir_or_data: Optional[Union[str, Path, Dict, ray.ObjectRef]],
+        storage_mode: CheckpointStorage,
+        checkpoint_id: Optional[int] = None,
+        metrics: Optional[Dict] = None,
+        node_ip: Optional[str] = None,
+    ):
+        self.dir_or_data = dir_or_data
+        self.id = checkpoint_id
+        self.storage_mode = storage_mode
+
+        self.metrics = metrics or {}
+        self.node_ip = node_ip or self.metrics.get(NODE_IP, None)
+
```

Comment:
Got it, thanks for the clarification on dir_or_data vs. storage_mode!

I get that if dir_or_data is a future, we don't know if it's an in-memory or persistent checkpoint. But in the case that it's not a future, then should we raise an error if the type of dir_or_data mismatches with storage_mode? Like if a path is passed in, but storage_mode is memory, then raise an error for example.

Reply:
For persistent checkpoints, `dir_or_data` will be a dict until `commit()` writes it to disk.
But I think for memory checkpoints we can enforce dicts or object refs. added a check